### PR TITLE
Upgrade to Ecto 3

### PR DESCRIPTION
**Pull request by wiserfirst**
1. Upgrade to Ecto 3
2. Upgrade to Elixir 1.7
3. Bump deps

### CODE REVIEW

File: lib/dal.ex [hunk 1]
```diff
@@ -329,17 +332,27 @@ defmodule DAL do
...
+  def discover(nil), do: nil
+
   def discover(queryable) do
     Discovery.fetch(queryable)
   end
 
   def preload(struct_or_structs_or_nil, preloads, opts \\ []) do
-    Ecto.Repo.Preloader.preload(struct_or_structs_or_nil, DAL, preloads, opts)
+    repo = discover(struct_or_structs_or_nil)
+    Ecto.Repo.Preloader.preload(struct_or_structs_or_nil, repo, preloads, opts)
```
Comment (coderdan) [inline lib/dal.ex]: What's the motivation for this change?

Comment (wiserfirst) [reply lib/dal.ex]: So we were passing in `DAL` to `Ecto.Repo.Preloader.preload/4` and in Ecto 2, it was calling `.all` on the input argument. But in Ecto 3, the implementation has changed. Now it expects this second argument to be an actual repo rather than a module name and it calls 
```
```

Comment (coderdan) [reply lib/dal.ex]: OK understood. The `DAL` should look like a Repo anyway but I think it's inconsequential.
